x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/TimeFunction.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
+
+import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateTimeProcessor.DateTimeExtractor;
+import org.elasticsearch.xpack.sql.expression.gen.processor.Processor;
+import org.elasticsearch.xpack.sql.tree.Source;
+
+import java.time.OffsetTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoField;
+
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isDateOrTime;
+import static org.elasticsearch.xpack.sql.util.DateUtils.asTimeAtZone;
+
+public abstract class TimeFunction extends DateTimeFunction {
+
+    TimeFunction(Source source, Expression field, ZoneId zoneId, DateTimeExtractor extractor) {
+        super(source, field, zoneId, extractor);
+    }
+
+    public static Integer dateTimeChrono(OffsetTime time, String tzId, String chronoName) {
+        return dateTimeChrono(asTimeAtZone(time, ZoneId.of(tzId)), ChronoField.valueOf(chronoName));
+    }
+
+    @Override
+    public Object fold() {
```

Comment:
Completely overriding this method here doesn't feel right. The field folding and the `ZonedDateTime` result is already handled in `BaseDateTimeFunction`...

Could this code be integrated in/moved to the `BaseDateTimeFunction` class and the concrete classes to implement only a `doFold` method? Maybe using a generic parameter for `doFold` depending on the class that implements it to be a date-type of class or a time-type of class?

`Object doFold(T dateOrTime)`

Reply:
I decided to do the refactoring mentioned here: #40566 and just delegate to the processor.